src/solvers/dgsem_tree/indicators_2d.jl

```diff
@@ -185,6 +185,100 @@ function (löhner::IndicatorLöhner)(u::AbstractArray{<:Any,4},
 end
 
 
+# this method is used when the indicator is constructed as for shock-capturing volume integrals
+function create_cache(indicator::Type{IndicatorIDP}, equations::AbstractEquations{2}, basis::LobattoLegendreBasis, number_bounds)
+  ContainerShockCapturingIndicator = Trixi.ContainerShockCapturingIndicatorIDP{real(basis)}(0, nnodes(basis), number_bounds)
+
+  cache = (; ContainerShockCapturingIndicator)
+
+  return cache
+end
+
+function (indicator_IDP::IndicatorIDP)(u::AbstractArray{<:Any,4}, semi, dg::DGSEM, t, dt; kwargs...)
+  @unpack alpha = indicator_IDP.cache.ContainerShockCapturingIndicator
+  alpha .= 0.0
+
+  indicator_IDP.IDPPositivity  &&
+    @trixi_timeit timer() "IDPPositivity"  IDP_positivity!(alpha, indicator_IDP, u, dt, semi)
+
+  # Calculate alpha1 and alpha2
+  @unpack alpha1, alpha2 = indicator_IDP.cache.ContainerShockCapturingIndicator
+  @threaded for element in eachelement(dg, semi.cache)
+    for j in eachnode(dg), i in 2:nnodes(dg)
+      alpha1[i, j, element] = max(alpha[i-1, j, element], alpha[i, j, element])
+    end
+    for j in 2:nnodes(dg), i in eachnode(dg)
+      alpha2[i, j, element] = max(alpha[i, j-1, element], alpha[i, j, element])
+    end
+    alpha1[1,            :, element] .= zero(eltype(alpha1))
+    alpha1[nnodes(dg)+1, :, element] .= zero(eltype(alpha1))
+    alpha2[:,            1, element] .= zero(eltype(alpha2))
+    alpha2[:, nnodes(dg)+1, element] .= zero(eltype(alpha2))
+  end
+
+  return nothing
+end
+
+@inline function IDP_positivity!(alpha, indicator_IDP, u, dt, semi)
+
+  # Conservative variables
+  for (index, variable) in enumerate(indicator_IDP.variables_cons)
+    IDP_positivity!(alpha, indicator_IDP, u, dt, semi, variable, index)
+  end
+
+  return nothing
+end
+
+@inline function IDP_positivity!(alpha, indicator_IDP, u, dt, semi, variable, index)
+  mesh, equations, dg, cache = mesh_equations_solver_cache(semi)
+  @unpack antidiffusive_flux1, antidiffusive_flux2 = cache.ContainerAntidiffusiveFlux2D
+  @unpack inverse_weights = dg.basis
+  @unpack positCorrFactor = indicator_IDP
+
+  @unpack var_bounds = indicator_IDP.cache.ContainerShockCapturingIndicator
+
+  var_min = var_bounds[index]
+
+  @threaded for element in eachelement(dg, semi.cache)
+    inverse_jacobian = cache.elements.inverse_jacobian[element]
+    for j in eachnode(dg), i in eachnode(dg)
+      var = variable(get_node_vars(u, equations, dg, i, j, element), equations)
+      if var < 0.0
+        println("Error: safe $variable is not safe. element=$element, node: $i $j, value=$var")
+      end
+
```

Comment:
```suggestion
```
This looks like a left-over debug statement?

Reply:
Actually, this is still very helpful. If you receive this warning, you know even the low-order solution would crash here and you have to lower the CFL number.

Reply:
But does it make sense to continue the solution in that case? That is, outside from pure development/debugging, do you ever continue the simulation if you see this warning? If not, I suggest to increase the severity of this check from a warning to an error, i.e., to stop the simulation.

Especially in production runs that go for a long time, a warning is easily overlooked.

Reply:
Yes, that's a good idea. Usually the simulation crashes shortly after that anyway.